src/calliope/model.py

```diff
@@ -341,26 +341,23 @@ def solve(self, force: bool = False, warmstart: bool = False, **kwargs) -> None:
         else:
             to_drop = []
 
-        kwargs["mode"] = self.config.build.applied_keyword_overrides.get(
-            "mode", self.config.build.mode
-        )
-
-        this_solve_config = self.config.update({"solve": kwargs}).solve
+        solve_config = self.config.update({"solve": kwargs}).solve
+        mode = self.config.build.mode
```

Comment:
This isn't valid. If a user updated the mode as a kwarg to `build` then `self.config.build.applied_keyword_overrides` will store that update, but `self.config.build` will not. This allows `self.config.build` to always be used to create a new temporary build config on each call to `build` (i.e., kwargs are temporary and so do not get stored in the core config, only in the separated kwarg dict). So, here you'd need to call `self.config.update(self.config.applied_keyword_overrides).build.mode`. Or we add a new config object method which returns the config with kwarg overrides applied automatically.

Reply:
@brynpickering This is an issue, yeah... keeping `config.build` to the initial setup complicates things a bunch, it seems.

I think I have two suggested solutions:

### Solution 1
Changing `model._is_built` from a bool to a string flag that holds the latest built mode, which is set `build()`

This way, if `solve` is ran:
- if `_is_built` is `None` -> throw an error
- if `_is_built` does not match `build.mode`: solve but throw a warning to notify that the mode was changed
- if `_is_built` matches `build.mode`: just solve normally

### Solution 2

Just updating build directly, changing `applied_keyword_overrides` to `first_configuration` (or a better name), which holds the first ever configuration, and adding a method that 'recovers' the initial values.

Reply:
I guess we don't want to give `mode` such a special position in things - it could be that other config elements need to be passed between the two stages in future. From that perspective, I wouldn't use solution 1. 

I like having `build.config` always reflecting the loaded, first config and then having the ad-hoc overrides living elsewhere. Then, every time you want to re-build / re-solve, you know what base config you are working with and don't accidentally expect some former kwarg config to still be in place. However, it's easily missed. The reason for not actually updating the pydantic model is that updating it in-place is a pain if you have frozen it, a feature I would keep (i.e., you don't allow direct updates to the config options - `config.solve.solver = "gurobi"` is not allowed). Instead, you get a copy of the model, and constantly updating the config object with a copy of itself is a bad idea.

Reply:
I am not sure what you mean by frozen in this case... I can easily alter the configuration by just writing to it:
![image](https://github.com/user-attachments/assets/ffb5a2bc-86f7-4bbe-adc6-b581baa192bd)

I'd argue that having the option to 'reset' the model configuration to the initial state is similarly simple to understand (although a change from the current norm). We could even add methods to save both configurations (latest and initial) when calling `model.save_netcdf()`. This also brings the additional advantage that the saved data and the 'loaded' configuration will always correspond.

No approach will save us from the fact that python just lets you modify stuff as you see fit. Here is the current config again, but with a wrongful value:
![image](https://github.com/user-attachments/assets/5550a7ef-9fe7-4bd1-b504-11fd758ca829)

Reply:
The question here is how to keep the current approach manageable and reduce future accidental bugs (as I did here...). I think we should shift the way we see the problem to 'let's ensure things are right when needed' rather than 'lets shield the config from everything'. 

`pydantic` allows us to easily re-run the validation whenever needed. But this only easily achievable if the config object is up-to-date:

![image](https://github.com/user-attachments/assets/38e8404b-9274-4883-867b-66b4385560d9)

Reply:
OK, I didn't actually freeze `build` and `solve` in the end! But you can't do `model.config.init.name = "new_name"` as there is a way to stop pydantic model attributes being set on-the-fly. I planned for this to extend to `build` and `solve`, it's just a matter of adding `"frozen": True` to the pydantic model config.

RE re-validating, that's what the new `update` method does. It applies new config and forces re-validation, returning the new pydantic model with the changes made. I had intended for this to be the only way to allow users to update the config (i.e., `model.config.build.mode = "give coconut"` would not be allowed), so that we get re-validation _every_ time. It's a runtime hit, but negligible given how small the pydantic model is.

The question then is whether we update in-place or not. [Returning an updated copy of the model is what pydantic recommends when models are otherwise frozen](https://docs.pydantic.dev/2.9/concepts/serialization/#model_copy). So, we can freeze it and shield the config by giving only one route for config updates which are _always_ re-validated. But we have a copy of the config returned to us rather than it updating in-place. This seems to be more the direction projects are going in (e.g. xarray has moved to returning copies of objects rather than allowing in-place operations). We could make the model with the current kwarg overrides applied more easily accessible with something like `model.config.with_kwargs_overrides`...?

Reply:
Oh! If freezing the config is possible, then avoiding in-place modifications is the way to go.

One last argument for making `model.config` reflect the current state rather than the initial state: I'd say it's more 'natural' to how most code / programs operate. I suspect most people will expect similar behaviour.

If someone does `model.config = model.config.update(my_kwargs)`, they'd most likely expect their changes to be reflected in `model.config`, rather than `model.config._kwargs` or `model.config.with_overrides`.

We can go either way, though. This is just minor preferences.

Reply:
Yeah, `model.config = model.config.update(my_kwargs)` would be feasible if we didn't freeze that level, but it isn't an _in-place_ operation (just `model.config.update(my_kwargs)` and expecting contents of config to be updated would be). But it feels safer to not allow `model.config` to be updated either (otherwise we have to validate that it is an appropriate type! What would stop you from doing `model.config = "foo"` other than the fact that model.config itself can be a frozen pydantic model?). 

I should say that I agree with you entirely that having to call/check `model.config.with_kwargs_overrides` is an annoyance that might cause issues down the line, so I'm keen to iterate to find a better solution to it!

Reply:
Ok. I think this will start taking form once the whole approach to `config` is set up.
Here is my proposal for now: I use `self.config.update(self.config.applied_keyword_overrides).build.mode` in this PR to get it in the other one.

Then we decide the final approach once the `CalliopeConfig` object is fully complete (also with techs, nodes, data tables, etc)?

Reply:
I already have `data_tables` up and going (including testing!) [here](https://github.com/calliope-project/calliope/tree/add-config-obj-replace-schemas). I think it's better to put a pin on this until the rest is finished, to make a better decision.

Reply:
Sounds good @irm-codebase . I suspect it will affect the config and the math. Both are places where we can update the dict via kwargs at different stages in the process. The data tables won't be so affected by it, I hope!

Reply:
This is how I think the configuration object would end up looking like:

```python
class CalliopeConfig(ConfigBaseModel):
    """Calliope configuration class."""
    model_config = {"title": "config"}
    init: Init = Init()
    build: Build = Build()
    solve: Solve = Solve()
    techs: dict[str, Tech]
    nodes: dict[str, Node]
    overrides: dict[str, Self] = None
    scenarios: dict[str, UniqueList[str]] = None
    data_tables: dict[str, DataTable] = None
```

I already made sure that `data_tables` are possible. I think the most complicated bits will be overrides, templates (which we'll most likely need to handle pre-pydantic) and some aspects of the math.